### PR TITLE
Add `TorchDynamo` callback

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Requires nightly PyTorch (1.13), `functorch`, and `torchdynamo`. The easiest way is to clone `torchdynamo` and run [`make setup_nightly`](https://github.com/pytorch/torchdynamo/blob/11873be7ec66b71cc065ca756bc6b5c292bef820/Makefile#L52-L56)

Resources:

- https://github.com/pytorch/torchdynamo
- How training is supported: https://dev-discuss.pytorch.org/t/torchdynamo-update-6-training-support-with-aotautograd/570
- Comparison of compiler projects: https://dev-discuss.pytorch.org/t/the-nuances-of-pytorch-graph-capture/501
- Inductor as the default backend: https://dev-discuss.pytorch.org/t/torchinductor-a-pytorch-native-compiler-with-define-by-run-ir-and-symbolic-shapes/747/2

Notes:
- I have written tests, but they won't run in CI because the dependency is not installed. Should it be? I lean towards no as it's under active development and there might be breaking changes.
- There is a single non-mock test with different backends that show multiple errors due to bugs/unsupported features in `torchdynamo`.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: src/pytorch_lightning/callbacks/torchdynamo.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        from pytorch_lightning.callbacks import TorchDynamo
+
+        # defaults to always use `"nvfuser"`
+        trainer = Trainer(callbacks=TorchDynamo())
+
+        # custom backend per stage
+        trainer = Trainer(callbacks=TorchDynamo({"train": "nvfuser", "predict": "fx2trt"})
+    """
+
+    def __init__(self, backend: Union[_BACKEND, Dict[str, _BACKEND]] = "nvfuser") -> None:
```
Comment (carmocca) [inline src/pytorch_lightning/callbacks/torchdynamo.py:42]: We might want to choose different backends per running stage by default.

File: src/pytorch_lightning/callbacks/torchdynamo.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+                raise ValueError("You passed an empty dictionary as the backend options.")
+            for key, value in backend.items():
+                if key not in list(RunningStage):
+                    stages = [stage.value for stage in list(RunningStage)]
+                    raise ValueError(f"The stage {key!r} should be one of {stages}.")
+                if isinstance(value, str):
+                    _check_valid_backend(value)
+
+        self.default_backend: str = "nvfuser" if is_cuda_available() else "nnc"
+        self._backend = backend
```
Comment (carmocca) [inline src/pytorch_lightning/callbacks/torchdynamo.py:65]: We might want to choose different backends per running stage by default.
